src/outputs/ascent.cpp

```diff
@@ -0,0 +1,203 @@
+  // the required fields cannot be resolved by Ascent.
+  // See https://ascent.readthedocs.io/en/latest/AscentAPI.html#field-filtering
+  // ascent_opts["field_filtering"] = "true";
+  ascent.open(ascent_opts);
+
+  // create root node for the whole mesh
+  Node root;
+
+  for (auto &pmb : pm->block_list) {
+    // create a unique id for this MeshBlock
+    const std::string &meshblock_name = "domain_" + std::to_string(pmb->gid);
+    Node &mesh = root[meshblock_name];
+
+    // add basic state info
+    mesh["state/domain_id"] = pmb->gid;
+    mesh["state/cycle"] = tm->ncycle;
+    mesh["state/time"] = tm->time;
+
+    auto &bounds = pmb->cellbounds;
+    auto ib = bounds.GetBoundsI(IndexDomain::entire);
+    auto jb = bounds.GetBoundsJ(IndexDomain::entire);
+    auto kb = bounds.GetBoundsK(IndexDomain::entire);
+    int nx = ib.e - ib.s + 1;
+    int ny = jb.e - jb.s + 1;
+    int nz = kb.e - kb.s + 1;
+    uint64_t ncells = nx * ny * nz;
+
+    auto ib_int = bounds.GetBoundsI(IndexDomain::interior);
+    auto jb_int = bounds.GetBoundsJ(IndexDomain::interior);
+    auto kb_int = bounds.GetBoundsK(IndexDomain::interior);
+
+    auto &coords = pmb->coords;
+    Real dx1 = coords.CellWidth<X1DIR>(ib.s, jb.s, kb.s);
+    Real dx2 = coords.CellWidth<X2DIR>(ib.s, jb.s, kb.s);
+    Real dx3 = coords.CellWidth<X3DIR>(ib.s, jb.s, kb.s);
+    std::array<Real, 3> corner = coords.GetXmin();
+
+    // create the coordinate set
+    mesh["coordsets/coords/type"] = "uniform";
+
+    mesh["coordsets/coords/dims/i"] = nx + 1;
+    mesh["coordsets/coords/dims/j"] = ny + 1;
+    if (nz > 1) {
+      mesh["coordsets/coords/dims/k"] = nz + 1;
+    }
+
+    // add origin and spacing to the coordset (optional)
+    mesh["coordsets/coords/origin/x"] = corner[0];
+    mesh["coordsets/coords/origin/y"] = corner[1];
+    if (nz > 1) {
+      mesh["coordsets/coords/origin/z"] = corner[2];
+    }
+
+    mesh["coordsets/coords/spacing/dx"] = dx1;
+    mesh["coordsets/coords/spacing/dy"] = dx2;
+    if (nz > 1) {
+      mesh["coordsets/coords/spacing/dz"] = dx3;
+    }
+
+    // add the topology
+    mesh["topologies/topo/type"] = "uniform";
+    mesh["topologies/topo/coordset"] = "coords";
+
+    // indicate ghost zones with ascent_ghosts set to 1
+    Node &n_field = mesh["fields/ascent_ghosts"];
+    n_field["association"] = "element";
+    n_field["topology"] = "topo";
+    n_field["values"].set(conduit::DataType::int32(ncells)); // NOLINT
+    conduit::int32_array vals_array = n_field["values"].value();
+
+    int idx = 0;
+    for (int k = kb.s; k <= kb.e; k++) {
+      for (int j = jb.s; j <= jb.e; j++) {
+        for (int i = ib.s; i <= ib.e; i++) {
+          if ((i < ib_int.s) || (ib_int.e < i) || (j < jb_int.s) || (jb_int.e < j) ||
+              ((nz > 1) && ((k < kb_int.s) || (kb_int.e < k)))) {
+            vals_array[idx] = 1;
+          } else {
+            vals_array[idx] = 0;
+          }
+          idx++;
+        }
+      }
+    }
```

Comment:
For large meshblocks doing this might be slow... I wonder if we should support doing this on device in the limit of large meshblocks, like NX1 ~ NX2 ~ NX3 ~ 100?

(No need to do this now. I'm just thinking out loud.)

Reply:
Good point.
Now that you mention it I think there's an even better solution.
Given that our meshblock size is fixed, I can just create one object tied to the class and then reuse every time the function is called.
I'll update this.

Reply:
:+1: Nice solution.